File: src/cpp/AdHoc.cpp
//: src/cpp/AdHoc.cpp
// All C++ examples tested with https://cpp.sh
#include <string>
#include <iostream>
using namespace std;

class Person {
    public:
    void eat() { cout << "eating pizza" << endl; }
};

class Robot {
    public:
    void charge() { cout << "charging" << endl; }
};

void nourish(Person p) { p.eat(); }

void nourish(Robot r) { r.charge(); }

int main() {
    nourish(Person());
    nourish(Robot());
}

File: src/java/src/AdHoc.java
//: src/java/src/AdHoc.java
class Person {
  void consume() { System.out.println("eating pizza"); }
}

class Robot {
  void charge() { System.out.println("charging"); }
}

public class AdHoc { // Overloading:
  static void nourish(Person subject) {
    subject.consume();
  }
  static void nourish(Robot subject) {
    subject.charge();
  }
  public static void main(String... args) {
    nourish(new Person());  // Determined at compile time
    nourish(new Robot());
  }
}